### PR TITLE
Add fixture loading/unloading functionality and usable fixture files

### LINKED ISSUE #11: [Enchancenent] OpenSearch test fixture loader and unloader
**Is your feature request related to a problem? Please describe.**

Monterey currently does not have any method of loading indices and data from a file to OpenSearch, which limits the ability for functional test developers to setup custom test environments. It also lacks a method of unloading data from OpenSearch, making it difficult to tear down the test environment and return OpenSearch to a clean state. The lack of these two features increases the difficulty of porting over existing OpenSearch Dashboards tests that require specific test environments and constrains the testing capabilities of the Monterey framework.

**Describe the solution you'd like**

Monterey should be able to parse the contents of a file containing JSONs that represent indices and data elements, and load and unload this parsed data through the use of the OpenSearch REST API. Ideally, this should be implemented as Cypress custom commands using the Cypress `readFile` and `request` functions to smoothly integrate the loading and unloading functionality into the existing Monterey framework.

**Describe alternatives you've considered**

One alternative is to directly use the `osd-opensearch-archiver` package, which is the existing data file loading and unloading implementation in the OpenSearch Dashboards testing framework. While utilizing this package would have the advantage of providing a solution that is already tailored to Monterey's specific needs, `osd-opensearch-archiver` has dependencies on another OpenSearch Dashboards package as well as on an Elasticsearch package, which in turn have dependencies on other packages. Handling the dependencies for these packages would increase the overhead for maintaining the Monterey framework, as well as generally increase the complexity of the project; thus, I find the `osd-opensearch-archiver` package approach to be non-ideal if there exists a simpler solution, especially for the initial proof of concept stages of Monterey. 

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by avivbenchorin**
### Description

Currently, the `dashboard_filtering_spec.js` functional test and the Monterey framework in general do not have a method of setting up and tearing down test environments in OpenSearch (as is described in #11). This PR aims to address this issue by adding custom Cypress commands that handle reading index and data JSONs from a fixture file, and subsequently interacting with OpenSearch through the OpenSearch REST API. 

The three added functions are:
1. `importJSONMapping` - read in and parse a fixture file containing indices, and then make `PUT` requests to add the indices and their mappings through the OpenSearch [create index API](https://opensearch.org/docs/opensearch/rest-api/create-index/).
2.  `clearJSONMapping` - read in and parse a fixture file containing indices, and then make `DELETE` requests to remove the indices through the OpenSearch [delete index API](https://www.elastic.co/guide/en/elasticsearch/reference/7.10/indices-delete-index.html).
3. `importJSONDoc` - read in and parse a fixture file containing doc data JSONs, and then make `POST` requests to import the data through the OpenSearch [bulk API](https://opensearch.org/docs/opensearch/rest-api/document-apis/bulk/). This function imports the doc JSONs across multiple requests rather than only one to avoid putting OpenSearch under memory pressure (read more [here](https://www.elastic.co/guide/en/elasticsearch/reference/master/tune-for-indexing-speed.html#_use_bulk_requests)). This function can only consistently work with files that are smaller than 20-25MB on the Electron browser, but can handle larger files on Firefox and Chrome.

The fixture files that can used with these functions are located in the `cypress/fixtures/dashboard` directory, with the `opensearch-dashboards` directory containing information related to the `.kibana_1` index and saved visualizations and the `data` directory containing other test-specific indices and data. The fixtures titled `mappings.json.txt` contain JSONs representing the specifications of indices, while the fixtures titled `data.json.txt` contain JSONs representing data to populate the aforementioned indices. Whenever importing indices and their associated data, always import the indices first and then the data second to prevent any issues. 

This PR also incorporates these functions into the `dashboard_filtering_spec.js` test in the `before` and `after` hooks that run at the start and end of the entire test run. Be sure NOT to run the `clearJSONMapping` function on the `opensearch-dashboards/mappings.json.txt` file, as the current functionality is unable to properly re-import and migrate the `.kibana_1` index and you may be unable to run the test file successfully in the future.

The `dashboard_filtering_spec.js` test file can now be run successfully without external test environment setup in Chrome and Firefox, but fails in the Electron browser when trying to import the large `data.json.txt` in `cypress/fixtures/dashboard/opensearch-dashboards` to OpenSearch.

Signed-off-by: Aviv Benchorin <benchori@amazon.com>

### Test

To run the `dashboard_filtering_spec.js` test, first make sure you have Cypress installed (instructions can be found [here](https://docs.cypress.io/guides/getting-started/installing-cypress)). Afterwards, you can follow these instructions for either using headless mode or using the Cypress Test Runner:

#### Headless

From the Monterey project root directory, run `npx cypress run  --browser <browser name> --headless --spec cypress/integration/dashboard/dashboard_filtering_spec.js`. This test has been tested using `chrome` and `firefox` as the browser and ran successfully, while using `electron` will fail. Below is an example of the results report given after a successful run of the test:

image

#### Cypress Test Runner

1. From the Monterey project root directory, run `npx cypress open`, the window shown below should open up.

image

2. Click on the `dashboard_filtering_spec.js` test, and the Cypress Test Runner will open, starting to run the test. Below are two halves of a successful run of the test:

https://user-images.githubusercontent.com/47436232/127723173-3d0dd1a3-91d7-42ed-baeb-4b44bceab7b7.mp4

https://user-images.githubusercontent.com/47436232/127723175-2cb76c7b-349b-4a57-a60c-4c27bc2cdcb0.mp4

 

### Issues Resolved

Partially resolves: #11

### Check List

- [ ] New functionality includes testing. 
    - [x] All tests pass
- [ ] New functionality has been documented. 
    - [x] New functionality has javadoc added 
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -32,5 +32,6 @@
     "eslint-plugin-promise": "^5.1.0",
     "eslint-plugin-react": "^7.24.0",
     "typescript": "^4.3.5"
-  }
+  },
+  "dependencies": {}
```
Comment (tianleh) [inline package.json]: Is this intentional? If not, let's remove it.

File: cypress/support/commands.js [hunk 2]
```diff
@@ -250,8 +248,102 @@ Cypress.Commands.add('checkElementComponentContainsValue', (mainSelector, compon
...
+      const body = { "index": {"_id": id, "_index": index }  }
+      const oneLineBody = JSON.stringify(body).replace('\n', '')
+      const oneLineSource = JSON.stringify(source).replace('\n', '')
+      const oneLineDoc = oneLineBody + '\n' + oneLineSource
+      bulkLines[0].push(oneLineDoc)
+      
+      line = line + 1
+      if(line % bulkMax === 0) {
+        console.log(bucket)
+        console.log("Before response" + bucket)
```
Comment (tianleh) [inline cypress/support/commands.js]: Remove debugging statements for development.

**Comment by tianleh** [comment]
Since now we have an end-to-end test experience, can you add a section in the description called "Test" which will list the test steps/commands and also sample snippet from a successful Cypress run?

File: cypress/support/commands.js [hunk 3]
```diff
@@ -250,8 +248,102 @@ Cypress.Commands.add('checkElementComponentContainsValue', (mainSelector, compon
...
+})
+
+/**
+ * Read indices from a file and send to OpenSearch using the create index API
+ * @param {String} filename File path (with its root at the directory containing the cypress.json file)
+ * @param {String} hostname Host name for OpenSearch
+ * @param {String} port Port for OpenSearch
+ */
+
+Cypress.Commands.add('importJSONMapping', (filepath, hostname = 'localhost', port = '9200') => {
```
Comment (tianleh) [inline cypress/support/commands.js:264]: Nonblocking: we can later refactor these loading/unloading commands to their own file, say data_commands.js or json_commands.js

File: cypress/integration/dashboard/dashboard_filtering_spec.js [hunk 4]
```diff
@@ -12,234 +12,256 @@
...
+
+    cy.importJSONDoc('cypress/fixtures/dashboard/opensearch_dashboards/data.json.txt')
+    cy.importJSONDoc('cypress/fixtures/dashboard/data/data.json.txt')
   })
 
   after(() => {
-    // TO DO: Tear-down the indices and test data (see https://github.com/AvivBenchorin/monterey/issues/11)
-    // cy.request('DELETE', 'localhost:9200')
+    // WORK IN PROGRESS: https://github.com/AvivBenchorin/monterey/issues/11
+    cy.clearJSONMapping('cypress/fixtures/dashboard/data/mappings.json.txt')
```
Comment (tianleh) [inline cypress/integration/dashboard/dashboard_filtering_spec.js:27]: Shall the docs be cleaned as well?

Comment (avivbenchorin) [reply cypress/integration/dashboard/dashboard_filtering_spec.js:27]: In the original `dashboard_filtering.js` test, only the indices had `DELETE` requests sent to OpenSearch (the rabbit-hole for finding this functionality starts [here](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/9a2d3e6918760c4bba8b44b9e8dfcd14e459a916/test/functional/apps/dashboard/index.js#L43) and ends [here](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/9a2d3e6918760c4bba8b44b9e8dfcd14e459a916/packages/osd-opensearch-archiver/src/lib/indices/delete_index.ts#L41)). My understanding is that deleting the OpenSearch index removes all of the data associated with it, and thus deleting doc by doc is unnecessary ([this](https://discuss.elastic.co/t/deleting-documents-vs-deleting-the-whole-index-and-rebuilding-it-periodically/127018) thread also suggests that deleting an index is more efficient than manually removing documents).

File: package.json (hunk 1 above)
Comment (avivbenchorin) [reply package.json]: I think it was a remnant from experimenting with `node-fetch`which I had installed using `npm`. I removed it now.

File: cypress/support/commands.js (hunk 2 above)
Comment (avivbenchorin) [reply cypress/support/commands.js]: The statements have been cut.

File: cypress/support/commands.js (hunk 3 above)
Comment (avivbenchorin) [reply cypress/support/commands.js:264]: I agree, the next PR can address splitting the `commands.js` file into multiple, more descriptive files.

**Comment by avivbenchorin** [comment]
> Since now we have an end-to-end test experience, can you add a section in the description called "Test" which will list the test steps/commands and also sample snippet from a successful Cypress run?

A test section has been added!

**Review by tianleh — APPROVED** [review]

**Comment by riysaxen-amzn** [comment]
> > Since now we have an end-to-end test experience, can you add a section in the description called "Test" which will list the test steps/commands and also sample snippet from a successful Cypress run?
> 
> A test section has been added!

great

**Review by riysaxen-amzn — APPROVED** [review]

**Pull request merged**